neqo-transport/src/recovery.rs

```diff
@@ -209,12 +227,152 @@ impl LossRecoverySpaces {
+            qinfo!([self], "slow start; cwnd {}", self.congestion_window);
+        } else {
+            // Congestion avoidance.
+            qinfo!(
+                [self],
+                "congestion avoidance; cwnd {} ssthresh {}",
+                self.congestion_window,
+                self.ssthresh
+            );
+            self.congestion_window += u64::try_from(MAX_DATAGRAM_SIZE * pkt.size as usize).unwrap()
+                / self.congestion_window
+        }
+    }
+
+    fn on_packets_lost(
+        &mut self,
+        now: Instant,
+        largest_acked_sent: Option<Instant>,
+        pto: Duration,
+        lost_packets: &[SentPacket],
+    ) {
+        for pkt in lost_packets {
+            assert!(self.bytes_in_flight >= u64::try_from(pkt.size).unwrap());
+            self.bytes_in_flight -= u64::try_from(pkt.size).unwrap();
+        }
+
+        qdebug!([self], "Pkts lost {}", lost_packets.len());
+
+        let last_lost_pkt = lost_packets.last().unwrap();
+        self.on_congestion_event(now, last_lost_pkt.time_sent);
+
+        let in_persistent_congestion = {
+            let congestion_period = pto * PERSISTENT_CONG_THRESH;
+            largest_acked_sent < Some(last_lost_pkt.time_sent - congestion_period)
+        };
+        if in_persistent_congestion {
+            qinfo!([self], "persistent congestion");
+            self.congestion_window = MIN_CONG_WINDOW;
+        }
+    }
+
+    fn on_packet_sent(&mut self, size: usize) {
+        self.bytes_in_flight += u64::try_from(size).unwrap();
+        qdebug!(
+            [self],
+            "Pkt Sent len {}, bif {}, cwnd {}",
+            size,
+            self.bytes_in_flight,
+            self.congestion_window
+        );
+        assert!(self.bytes_in_flight <= self.congestion_window);
+    }
+
+    fn in_congestion_recovery(&self, sent_time: Instant) -> bool {
+        self.congestion_recovery_start_time
+            .map(|start| sent_time <= start)
+            .unwrap_or(false)
+    }
+
+    fn on_congestion_event(&mut self, now: Instant, sent_time: Instant) {
+        // Start a new congestion event if packet was sent after the
+        // start of the previous congestion recovery period.
+        if !self.in_congestion_recovery(sent_time) {
+            self.congestion_recovery_start_time = Some(now);
+            self.congestion_window /= 2; // kLossReductionFactor = 0.5
+            self.congestion_window = max(self.congestion_window, MIN_CONG_WINDOW);
+            self.ssthresh = self.congestion_window;
+            qinfo!(
+                [self],
+                "Cong event -> recovery; cwnd {}, ssthresh {}",
+                self.congestion_window,
+                self.ssthresh
+            );
+        } else {
+            qdebug!([self], "Cong event but already in recovery");
+        }
+    }
+
+    fn app_limited(&self) -> bool {
+        //TODO(agrover): how do we get this info??
```

Comment:
This is tricky, but the basic idea is that you want to avoid congestion control increases if you are unable to send a full congestion window.  I don't think that we've settled on this in the spec, so leaving this as a TODO is fine.  Make an issue.

Ideas.

We are app limited if:

* there is no application data queued for sending (that's weak, but it might be enough)
* bytes_in_flight never reached cwnd in the last RTT (that might be too strong and cause us to never increase, particularly when you consider pacing, so you might have to pad the time or reduce the target somehow)

Ian had a few other ideas, but they are all hard to execute.  Defer this.  False is a fine starting position.

Reply:
#350